skrub/_agg_joiner.py

```diff
@@ -6,49 +6,140 @@
 table with the main table.
 """
 
+from itertools import product
+
 import numpy as np
+import pandas as pd
 from sklearn.base import BaseEstimator, TransformerMixin
 from sklearn.utils.multiclass import type_of_target
 from sklearn.utils.validation import check_is_fitted
 
 from skrub import _dataframe as sbd
 from skrub import _join_utils
+from skrub import _selectors as s
 from skrub._dataframe._namespace import get_df_namespace, is_pandas, is_polars
-from skrub._dataframe._pandas import _parse_argument
+from skrub._dispatch import dispatch
 from skrub._utils import atleast_1d_or_none
 
 from ._check_input import CheckInputDataFrame
 
-NUM_OPERATIONS = ["sum", "mean", "std", "min", "max", "hist", "value_counts"]
-CATEG_OPERATIONS = ["mode", "count", "value_counts"]
-ALL_OPS = NUM_OPERATIONS + CATEG_OPERATIONS
+try:
+    import polars as pl
+except ImportError:
+    pass
+
+SUPPORTED_OPS = ["count", "mode", "min", "max", "sum", "median", "mean", "std"]
+
 
+def aggregate(table, key, cols_to_agg, operations, suffix):
+    """Aggregate `table` on `key` and compute statistics on `cols_to_agg`.
 
-def split_num_categ_operations(operations):
-    """Separate aggregator operators input by their type.
+    Operations ["sum", "median", "mean", "std"] are only supported for numeric columns.
+
+    Add a suffix to columns not in `key` and sort columns by name.
 
     Parameters
     ----------
-    operations : list of str
-        The input operators names.
+    table : DataFrame
+        The input dataframe to aggregate.
+
+    key : str or iterable of str
+        The columns used as keys to aggregate on.
+
+    cols_to_agg : str or iterable of str
+        The columns to aggregate.
+
+    operations : str or iterable of str
+        The reduction functions to apply on columns
+        in ``cols_to_agg`` during the aggregation.
+
+    suffix : str
+        The suffix appended to output columns. Will only be applied
+        to columns created by the aggregations.
 
     Returns
     -------
-    num_operations, categ_operations : Tuple of list of str
-        List of operator names.
+    DataFrame
+        The aggregated output.
     """
-    num_operations, categ_operations = [], []
-    for operation in operations:
-        # hist(5) -> hist
-        op_root, _ = _parse_argument(operation)
-        if op_root in NUM_OPERATIONS:
-            num_operations.append(operation)
-        if op_root in CATEG_OPERATIONS:
-            categ_operations.append(operation)
-        if op_root not in ALL_OPS:
-            raise ValueError(f"operations options are {ALL_OPS}, got: {operation=!r}.")
 
-    return num_operations, categ_operations
+    # summing strings works in pandas, not in polars
+    num_only_operations = ["sum", "median", "mean", "std"]
+
+    key = atleast_1d_or_none(key)
+    cols_to_agg = atleast_1d_or_none(cols_to_agg)
+    operations = atleast_1d_or_none(operations)
```

Comment:
All of these checks have been performed upstream in `check_input` method. We should probably clarify our intent with this function; should it only be called by `AggJoiner` and `AggTarget`, or should it be publicly accessible?

Reply:
You're right, they're also leftovers from the original aggregate => removing them

I think `aggregate` should only be called by `AggJoiner` and `AggTarget`, users doing groupbys are most likely to directly use pandas / polars directly IMO

Reply:
The checks are now left the the Joiners, I enforced that the parameters are iterables when they arrive to `aggregate`